DQM/L1TMonitor/src/L1MUTK.cc

```diff
@@ -0,0 +1,238 @@
+    cout << "L1T Monitoring histograms will be saved to " << outputFile_.c_str() << endl;
+  }
+
+  bool disable = ps.getUntrackedParameter<bool>("disableROOToutput", false);
+  if(disable){
+    outputFile_="";
+  }
+
+  // input tags
+  candInputTag_ = ps.getParameter<edm::InputTag>("candInputTag");
+  LogInfo("L1MUTK") << " L1MUTK::L1MUTK candInputTag " << candInputTag_;
+
+  if ( dbe !=NULL ) {
+    dbe->setCurrentFolder("L1T/L1MUTK");
+  }
+
+}
+
+L1MUTK::~L1MUTK()
+{
+}
+
+void L1MUTK::beginJob()
+{
+  nev_ = 0;
+}
+
+void L1MUTK::beginRun(const edm::Run& r, const edm::EventSetup& c)
+{
+  if(nev_==0) {
+      book_(c);
+  }
+}
+
+void L1MUTK::endJob(void)
+{
+  if(verbose_) cout << "L1MUTK: end job...." << endl;
+  LogInfo("EndJob") << "analyzed " << nev_ << " events"; 
+
+ if ( outputFile_.size() != 0  && dbe ) dbe->save(outputFile_);
+
+ return;
+}
+
+void L1MUTK::analyze(const Event& e, const EventSetup& c)
+{
+  
+  nev_++; 
+
+  edm::Handle<CandView> aH; 
+  e.getByLabel(candInputTag_, aH);
+
+  if(!aH.isValid()){
+    LogDebug("L1MUTK") << " ::analyze Can't find CandView with label **candInputTag**";
+  }
+
+  const CandView& cands(*aH.product());
+
+  int track = 0, trackPt=0;
+  double maxPt = 0.;
+  double dxy;
+  for (auto cand : cands){
+    track++;
+    if (cand.pt() >= maxPt){
+      trackPt = track;
+      maxPt = cand.pt();
+    }
+  }
+
+  track = 0;
+  for (auto cand : cands){
+    track++;
+    if ( track == trackPt ){
+      dxy = cand.vx()*sin(cand.phi())-cand.vy()*cos(cand.phi());
+
+      LogDebug("L1MUTK") << " ::analyze track " 
+          << track << " Pt " << cand.pt() << " Phi " << cand.phi() << " Eta " << cand.eta() << " Dxy " << dxy 
+          << " Z " << cand.vz() << " Q " << cand.charge() << " cand.pdgId " << cand.pdgId();
+
+      gem_Pt->Fill(cand.pt());
+      gem_Phi->Fill(cand.phi());
+      gem_Eta->Fill(cand.eta());
+      gem_Dxy->Fill(dxy);
+      gem_Z->Fill(cand.vz());
+      gem_Q->Fill(float(cand.charge()));
+      gem_Pdgid->Fill(float(cand.pdgId()));
+ 
+      if ( std::abs(cand.eta()) <= 1.1 ){
+	// Barrel
+        gem_Pt_barrel->Fill(cand.pt());
+        gem_Phi_barrel->Fill(cand.phi());
+        gem_Eta_barrel->Fill(cand.eta());
+        gem_Dxy_barrel->Fill(dxy);
+        gem_Z_barrel->Fill(cand.vz());
+        gem_Q_barrel->Fill(float(cand.charge()));
+        gem_Pdgid_barrel->Fill(float(cand.pdgId()));
+      } else {
+        // Endcap
+        gem_Pt_endcap->Fill(cand.pt());
+        gem_Phi_endcap->Fill(cand.phi());
+        gem_Eta_endcap->Fill(cand.eta());
+        gem_Dxy_endcap->Fill(dxy);
+        gem_Z_endcap->Fill(cand.vz());
+        gem_Q_endcap->Fill(float(cand.charge()));
+        gem_Pdgid_endcap->Fill(float(cand.pdgId()));
+      }
+    }
+  }
+}
+
+void L1MUTK::book_(const EventSetup& c)
+{
+  // get hold of back-end interface
+  DQMStore* dbe = 0;
+  dbe = Service<DQMStore>().operator->();
+
+  if ( dbe ) {
+    dbe->setCurrentFolder("L1T/L1MUTK");
```

Comment:
Hi Slava,

I put all other modifications which you suggest but that part I didn't 
understand.
Do we need something like:

void L1TkMuonMonitor::book_(const EventSetup& c, const DQMStore& dbe)  ?

SB

> In DQM/L1TMonitor/src/L1MUTK.cc 
> <https://github.com/slava77/cmssw/pull/69#discussion_r33595315>:
>
> > +        gem_Z_endcap->Fill(cand.vz());
> > +        gem_Q_endcap->Fill(float(cand.charge()));
> > +        gem_Pdgid_endcap->Fill(float(cand.pdgId()));
> > +      }
> > +    }
> > +  }
> > +}
> > +
> > +void L1MUTK::book_(const EventSetup& c)
> > +{
> > +  // get hold of back-end interface
> > +  DQMStore* dbe = 0;
> > +  dbe = Service<DQMStore>().operator->();
> > +
> > +  if ( dbe ) {
> > +    dbe->setCurrentFolder("L1T/L1MUTK");
>
> define a constant for the folder name so that it's not copy-pasted 
> multiple times
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/slava77/cmssw/pull/69/files#r33595315>.
>